File: python/code_13.py
import numpy as np
import re
import matplotlib.pyplot as plt 

inp = open('./data/dt_13.txt').read().split('\n')
### Folding & points coordinates
fold_info, idx = [], []
for x in inp :
    if "fold" in x :
        fold_info.append(x)
    elif x != '':
        idx.append(x.split(','))
idx = np.array(idx, dtype=int)

### paper template ###
# nummber of row and cols (i.e. paper dim) should be odd
r = 1 if max(idx[:, 1]) % 2 == 0 else 2 
c = 1 if max(idx[:, 0]) % 2 == 0 else 2 
paper = np.zeros(shape = (max(idx[:, 1]) + r, max(idx[:, 0]) + c), dtype=bool)
paper[(idx[:, 1], idx[:, 0])] = True

fold_xy = [int(re.sub('fold along .=(\\d+)$', '\\1', x)) for x in fold_info]
fold_nm = [re.sub('fold along (.)=\\d+$', '\\1', x) for x in fold_info]

### Folding paper
for i, nm in zip(fold_xy, fold_nm):
    if nm == 'x':
        paper = paper[:, :i] | np.flip(paper[:, (i+1):], 1)
    else:
        paper = paper[:i, :] | np.flip(paper[(i+1):, :], 0)
    
    if i == fold_xy[0]:
        print(f'Answer 1: {np.sum(paper)}')
### Part 2: code on image
plt.imshow(paper)
plt.show()

File: js/code_13.js
const fs = require('fs')
const [input1, input2] = fs
                .readFileSync('./data/dt_13.txt', 'utf8')
                .trim()
                .replace(/\r/g, "") //windows specific
                .split("\n\n");

/* url https://adventofcode.com/2021/day/13/input
code can be executed in browser's console, just use code line provided below for data extraction/read (not lines above)
const [input1, input2] = document.querySelector("body > pre").innerText.trim().split('\n\n'); */

//dot coordinates
const idx = input1.trim().split('\n').map((x) => {
    return x.split(',').map(Number);
});

const foldInfo = input2.trim().split('\n').map((x) => x
                       .match(/fold along (?<axis>[xy])=(?<position>\d+)/).groups)
                       .map((x) => [x.axis, Number(x.position)]);

const maxX = Math.max(...idx.map((p) => p[0])) + 1;
const maxY = Math.max(...idx.map((p) => p[1])) + 2;
let paper = [...Array(maxY)].map(a => Array(maxX).fill(' '));

//populating sheet of paper with dots
for (ix of idx) {
    paper[ix[1]][ix[0]] = '#';
};

//Folding sheet of paper
for (fold of foldInfo){

    if (fold[0] === 'x'){

        const firstHalf  = paper.map(a => a.slice(0, fold[1]));
		const secondHalf = paper.map(a => a.slice(-fold[1]));
        for (let y in secondHalf) {
            for (let x in secondHalf[y]) {
                let deltaX = fold[1] - parseInt(x) -1 ;
                firstHalf[y][x] = (firstHalf[y][x]!='#')?secondHalf[y][deltaX]:'#';
            };
        };
        paper = firstHalf;

    } else if (fold[0] === 'y') {

        const firstHalf  = paper.slice(0, fold[1]);
		const secondHalf = paper.slice(-fold[1]);
		for (let y in secondHalf) {
			for (let x in secondHalf[y]) {
				let deltaY = fold[1] - parseInt(y) - 1;
				firstHalf[y][x] = (firstHalf[y][x]!='#')?secondHalf[deltaY][x]:'#';
			};
		};
        paper = firstHalf;
    }
    if (fold[0] === foldInfo[0][0] && fold[1] === foldInfo[0][1]) {
        console.log('Answer 1: ', paper.flat().filter(a => a === '#').length);
    };
};

console.log('\nAnswer 2: \n');
console.log(paper.map(a => a.join("")).join('\n'));